File: go/internal/tools/builtins.go
package tools

import (
	"context"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"sort"
	"strings"
	"time"
)

const defaultTimeout = 30

// BuiltinDef defines a built-in tool's metadata and execution function. Name
// and Description appear in the Anthropic tool schema sent to the LLM,
// InputSchema is the JSON Schema object describing accepted parameters, and
// ExecuteFn is the callback invoked at runtime with the parsed input.
type BuiltinDef struct {
	Name        string
	Description string
	InputSchema map[string]any
	ExecuteFn   ToolFunc
}

func bashTool(_ context.Context, input map[string]any) ToolResult {
	command, _ := input["command"].(string)
	timeout := defaultTimeout
	if t, ok := input["timeout"].(float64); ok {
		timeout = int(t)
	}

	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(timeout)*time.Second)
	defer cancel()

	cmd := exec.CommandContext(ctx, "sh", "-c", command)
	out, err := cmd.CombinedOutput()
	output := strings.TrimSpace(string(out))

	if ctx.Err() == context.DeadlineExceeded {
		return ToolResult{
			Output:  fmt.Sprintf("Command timed out after %d seconds", timeout),
			IsError: true,
		}
	}

	isError := err != nil
	if output == "" {
		if isError {
			output = fmt.Sprintf("Command failed with exit code %d", cmd.ProcessState.ExitCode())
		} else {
			output = "(no output)"
		}
	}

	return ToolResult{Output: output, IsError: isError}
}

func readFileTool(_ context.Context, input map[string]any) ToolResult {
	path, _ := input["path"].(string)
	offset := 0
	limit := 2000
	if o, ok := input["offset"].(float64); ok {
		offset = int(o)
	}
	if l, ok := input["limit"].(float64); ok {
		limit = int(l)
	}

	data, err := os.ReadFile(path)
	if err != nil {
		return ToolResult{Output: err.Error(), IsError: true}
	}

	lines := strings.Split(string(data), "\n")
	if len(lines) > 0 && lines[len(lines)-1] == "" {
		lines = lines[:len(lines)-1]
	}

	end := offset + limit
	if end > len(lines) {
		end = len(lines)
	}
	selected := lines[offset:end]

	var numbered []string
	for i, line := range selected {
		numbered = append(numbered, fmt.Sprintf("%d\t%s", i+offset+1, line))
	}
	output := strings.Join(numbered, "\n")
	if offset+limit < len(lines) {
		output += fmt.Sprintf("\n... (%d more lines)", len(lines)-offset-limit)
	}

	return ToolResult{Output: output}
}

func writeFileTool(_ context.Context, input map[string]any) ToolResult {
	path, _ := input["path"].(string)
	content, _ := input["content"].(string)

	if err := os.WriteFile(path, []byte(content), 0o644); err != nil {
		return ToolResult{Output: err.Error(), IsError: true}
	}
	return ToolResult{Output: fmt.Sprintf("Written %d bytes to %s", len(content), path)}
}

func editFileTool(_ context.Context, input map[string]any) ToolResult {
	path, _ := input["path"].(string)
	oldText, _ := input["old_text"].(string)
	newText, _ := input["new_text"].(string)
	replaceAll, _ := input["replace_all"].(bool)

	data, err := os.ReadFile(path)
	if err != nil {
		return ToolResult{Output: err.Error(), IsError: true}
	}

	content := string(data)
	count := strings.Count(content, oldText)

	if count == 0 {
		return ToolResult{Output: "old_text not found in file", IsError: true}
	}
	if count > 1 && !replaceAll {
		return ToolResult{
			Output:  fmt.Sprintf("old_text found %d times — set replace_all to replace all occurrences", count),
			IsError: true,
		}
	}

	var newContent string
	if replaceAll {
		newContent = strings.ReplaceAll(content, oldText, newText)
	} else {
		newContent = strings.Replace(content, oldText, newText, 1)
	}

	if err := os.WriteFile(path, []byte(newContent), 0o644); err != nil {
		return ToolResult{Output: err.Error(), IsError: true}
	}

	replacements := 1
	if replaceAll {
		replacements = count
	}
	return ToolResult{Output: fmt.Sprintf("Replaced %d occurrence(s) in %s", replacements, path)}
}

func listDirectoryTool(_ context.Context, input map[string]any) ToolResult {
	path := "."
	if p, ok := input["path"].(string); ok && p != "" {
		path = p
	}

	info, err := os.Stat(path)
	if err != nil {
		return ToolResult{Output: err.Error(), IsError: true}
	}
	if !info.IsDir() {
		return ToolResult{Output: fmt.Sprintf("Not a directory: %s", path), IsError: true}
	}

	entries, err := os.ReadDir(path)
	if err != nil {
		return ToolResult{Output: err.Error(), IsError: true}
	}

	sort.Slice(entries, func(i, j int) bool {
		iDir := entries[i].IsDir()
		jDir := entries[j].IsDir()
		if iDir != jDir {
			return iDir
		}
		return strings.ToLower(entries[i].Name()) < strings.ToLower(entries[j].Name())
	})

	var lines []string
	for _, e := range entries {
		prefix := "[FILE]"
		if e.IsDir() {
			prefix = "[DIR]"
		}
		lines = append(lines, fmt.Sprintf("%s %s", prefix, e.Name()))
	}

	if len(lines) == 0 {
		return ToolResult{Output: "(empty directory)"}
	}
	return ToolResult{Output: strings.Join(lines, "\n")}
}

func searchFilesTool(_ context.Context, input map[string]any) ToolResult {
	path, _ := input["path"].(string)
	pattern, _ := input["pattern"].(string)

	var matches []string
	err := filepath.Walk(path, func(p string, info os.FileInfo, err error) error {
		if err != nil {
			return nil
		}
		matched, err := filepath.Match(pattern, filepath.Base(p))
		if err != nil {
			return nil
		}
		if matched {
			matches = append(matches, p)
		}
		return nil
	})
	if err != nil {
		return ToolResult{Output: err.Error(), IsError: true}
	}

	sort.Strings(matches)
	if len(matches) == 0 {
		return ToolResult{Output: "No matches found"}
	}
	return ToolResult{Output: strings.Join(matches, "\n")}
}

// BuiltinRegistry maps tool names to their definitions for all built-in tools.
var BuiltinRegistry = map[string]BuiltinDef{
	"bash": {
		Name:        "bash",
		Description: "Execute a shell command and return its output.",
		InputSchema: map[string]any{
			"type": "object",
			"properties": map[string]any{
				"command": map[string]any{
					"type":        "string",
					"description": "The shell command to execute",
				},
				"timeout": map[string]any{
					"type":        "integer",
					"description": "Timeout in seconds (default 30)",
				},
			},
			"required": []string{"command"},
		},
		ExecuteFn: bashTool,
	},
	"read_file": {
		Name:        "read_file",
		Description: "Read the contents of a file with line numbers. Supports offset and limit for large files.",
		InputSchema: map[string]any{
			"type": "object",
			"properties": map[string]any{
				"path": map[string]any{
					"type":        "string",
					"description": "Path to the file to read",
				},
				"offset": map[string]any{
					"type":        "integer",
					"description": "Line number to start reading from (0-based, default 0)",
				},
				"limit": map[string]any{
					"type":        "integer",
					"description": "Maximum number of lines to read (default 2000)",
				},
			},
			"required": []string{"path"},
		},
		ExecuteFn: readFileTool,
	},
	"write_file": {
		Name:        "write_file",
		Description: "Write content to a file, creating it if it doesn't exist or overwriting if it does.",
		InputSchema: map[string]any{
			"type": "object",
			"properties": map[string]any{
				"path": map[string]any{
					"type":        "string",
					"description": "Path to the file to write",
				},
				"content": map[string]any{
					"type":        "string",
					"description": "Content to write to the file",
				},
			},
			"required": []string{"path", "content"},
		},
		ExecuteFn: writeFileTool,
	},
	"edit_file": {
		Name:        "edit_file",
		Description: "Replace text in a file. Finds old_text and replaces with new_text. Fails if old_text is not found or matches multiple times (unless replace_all is true).",
		InputSchema: map[string]any{
			"type": "object",
			"properties": map[string]any{
				"path": map[string]any{
					"type":        "string",
					"description": "Path to the file to edit",
				},
				"old_text": map[string]any{
					"type":        "string",
					"description": "The exact text to find and replace",
				},
				"new_text": map[string]any{
					"type":        "string",
					"description": "The text to replace it with",
				},
				"replace_all": map[string]any{
					"type":        "boolean",
					"description": "Replace all occurrences (default false)",
				},
			},
			"required": []string{"path", "old_text", "new_text"},
		},
		ExecuteFn: editFileTool,
	},
	"list_directory": {
		Name:        "list_directory",
		Description: "List the contents of a directory, showing [DIR] and [FILE] prefixes.",
		InputSchema: map[string]any{
			"type": "object",
			"properties": map[string]any{
				"path": map[string]any{
					"type":        "string",
					"description": "Path to the directory to list (default: current directory)",
				},
			},
		},
		ExecuteFn: listDirectoryTool,
	},
	"search_files": {
		Name:        "search_files",
		Description: "Recursively search for files matching a glob pattern.",
		InputSchema: map[string]any{
			"type": "object",
			"properties": map[string]any{
				"path": map[string]any{
					"type":        "string",
					"description": "Directory to search in",
				},
				"pattern": map[string]any{
					"type":        "string",
					"description": "Glob pattern to match (e.g. '*.py', '**/*.json')",
				},
			},
			"required": []string{"path", "pattern"},
		},
		ExecuteFn: searchFilesTool,
	},
}

File: src/agentlings/tools/builtins.py
"""Built-in tool implementations: bash shell and filesystem operations."""

from __future__ import annotations

import subprocess
from pathlib import Path
from typing import Any

from agentlings.tools.registry import ToolResult

DEFAULT_TIMEOUT = 30


# ---------------------------------------------------------------------------
# Tool implementations
# ---------------------------------------------------------------------------


def _make_bash(default_timeout: int = DEFAULT_TIMEOUT) -> callable:
    """Create a bash tool function with a configurable default timeout."""

    def _bash(command: str, timeout: int | None = None) -> ToolResult:
        effective_timeout = timeout if timeout is not None else default_timeout
        try:
            result = subprocess.run(
                command,
                shell=True,
                capture_output=True,
                text=True,
                timeout=effective_timeout,
            )
            output = result.stdout
            if result.stderr:
                output += result.stderr
            output = output.strip()
            is_error = result.returncode != 0
            if not output:
                output = (
                    f"Command failed with exit code {result.returncode}"
                    if is_error
                    else "(no output)"
                )
            return ToolResult(output=output, is_error=is_error)
        except subprocess.TimeoutExpired:
            return ToolResult(
                output=f"Command timed out after {effective_timeout} seconds",
                is_error=True,
            )

    return _bash


def _read_file(path: str, offset: int = 0, limit: int = 2000) -> ToolResult:
    """Read a text file and return its contents with line numbers.

    Args:
        path: Path to the file to read.
        offset: Zero-based line number to start reading from.
        limit: Maximum number of lines to return.

    Returns:
        Numbered lines of the file, with a truncation notice if applicable.
    """
    try:
        p = Path(path)
        lines = p.read_text(encoding="utf-8").splitlines()
        selected = lines[offset : offset + limit]
        numbered = [f"{i + offset + 1}\t{line}" for i, line in enumerate(selected)]
        output = "\n".join(numbered)
        if offset + limit < len(lines):
            output += f"\n... ({len(lines) - offset - limit} more lines)"
        return ToolResult(output=output)
    except (FileNotFoundError, PermissionError, IsADirectoryError) as e:
        return ToolResult(output=str(e), is_error=True)


def _write_file(path: str, content: str) -> ToolResult:
    """Write content to a file, creating or overwriting as needed.

    Args:
        path: Path to the file to write.
        content: The text content to write.
    """
    try:
        Path(path).write_text(content, encoding="utf-8")
        return ToolResult(output=f"Written {len(content)} bytes to {path}")
    except (PermissionError, OSError) as e:
        return ToolResult(output=str(e), is_error=True)


def _edit_file(
    path: str, old_text: str, new_text: str, replace_all: bool = False
) -> ToolResult:
    """Find and replace text in a file.

    Fails if ``old_text`` is not found. If it matches multiple times,
    fails unless ``replace_all`` is set.

    Args:
        path: Path to the file to edit.
        old_text: Exact text to find.
        new_text: Replacement text.
        replace_all: If true, replace all occurrences instead of requiring a unique match.
    """
    try:
        p = Path(path)
        content = p.read_text(encoding="utf-8")
        count = content.count(old_text)

        if count == 0:
            return ToolResult(output="old_text not found in file", is_error=True)
        if count > 1 and not replace_all:
            return ToolResult(
                output=f"old_text found {count} times — set replace_all to replace all occurrences",
                is_error=True,
            )

        if replace_all:
            new_content = content.replace(old_text, new_text)
        else:
            new_content = content.replace(old_text, new_text, 1)

        p.write_text(new_content, encoding="utf-8")
        replacements = count if replace_all else 1
        return ToolResult(output=f"Replaced {replacements} occurrence(s) in {path}")
    except (FileNotFoundError, PermissionError) as e:
        return ToolResult(output=str(e), is_error=True)


def _list_directory(path: str = ".") -> ToolResult:
    """List directory contents with ``[DIR]`` and ``[FILE]`` prefixes.

    Args:
        path: Path to the directory to list. Defaults to the current directory.
    """
    try:
        p = Path(path)
        if not p.is_dir():
            return ToolResult(output=f"Not a directory: {path}", is_error=True)

        entries = sorted(p.iterdir(), key=lambda e: (not e.is_dir(), e.name.lower()))
        lines = []
        for entry in entries:
            prefix = "[DIR]" if entry.is_dir() else "[FILE]"
            lines.append(f"{prefix} {entry.name}")

        return ToolResult(output="\n".join(lines) if lines else "(empty directory)")
    except (PermissionError, OSError) as e:
        return ToolResult(output=str(e), is_error=True)


def _search_files(path: str, pattern: str) -> ToolResult:
    """Recursively search for files matching a glob pattern.

    Args:
        path: Root directory to search from.
        pattern: Glob pattern to match (e.g. ``"*.py"``, ``"**/*.json"``).
    """
    try:
        p = Path(path)
        matches = sorted(str(m) for m in p.rglob(pattern))
        if not matches:
            return ToolResult(output="No matches found")
        return ToolResult(output="\n".join(matches))
    except (PermissionError, OSError) as e:
        return ToolResult(output=str(e), is_error=True)


# ---------------------------------------------------------------------------
# Registry mapping tool names to their definitions
# ---------------------------------------------------------------------------


def build_builtin_registry(bash_timeout: int = DEFAULT_TIMEOUT) -> dict[str, dict[str, Any]]:
    """Build the built-in tool registry with configurable defaults."""
    return {
    "bash": {
        "name": "bash",
        "description": "Execute a shell command and return its output.",
        "input_schema": {
            "type": "object",
            "properties": {
                "command": {
                    "type": "string",
                    "description": "The shell command to execute",
                },
                "timeout": {
                    "type": "integer",
                    "description": f"Timeout in seconds (default {bash_timeout})",
                },
            },
            "required": ["command"],
        },
        "execute_fn": _make_bash(bash_timeout),
    },
    "read_file": {
        "name": "read_file",
        "description": "Read the contents of a file with line numbers. Supports offset and limit for large files.",
        "input_schema": {
            "type": "object",
            "properties": {
                "path": {
                    "type": "string",
                    "description": "Path to the file to read",
                },
                "offset": {
                    "type": "integer",
                    "description": "Line number to start reading from (0-based, default 0)",
                },
                "limit": {
                    "type": "integer",
                    "description": "Maximum number of lines to read (default 2000)",
                },
            },
            "required": ["path"],
        },
        "execute_fn": _read_file,
    },
    "write_file": {
        "name": "write_file",
        "description": "Write content to a file, creating it if it doesn't exist or overwriting if it does.",
        "input_schema": {
            "type": "object",
            "properties": {
                "path": {
                    "type": "string",
                    "description": "Path to the file to write",
                },
                "content": {
                    "type": "string",
                    "description": "Content to write to the file",
                },
            },
            "required": ["path", "content"],
        },
        "execute_fn": _write_file,
    },
    "edit_file": {
        "name": "edit_file",
        "description": "Replace text in a file. Finds old_text and replaces with new_text. Fails if old_text is not found or matches multiple times (unless replace_all is true).",
        "input_schema": {
            "type": "object",
            "properties": {
                "path": {
                    "type": "string",
                    "description": "Path to the file to edit",
                },
                "old_text": {
                    "type": "string",
                    "description": "The exact text to find and replace",
                },
                "new_text": {
                    "type": "string",
                    "description": "The text to replace it with",
                },
                "replace_all": {
                    "type": "boolean",
                    "description": "Replace all occurrences (default false)",
                },
            },
            "required": ["path", "old_text", "new_text"],
        },
        "execute_fn": _edit_file,
    },
    "list_directory": {
        "name": "list_directory",
        "description": "List the contents of a directory, showing [DIR] and [FILE] prefixes.",
        "input_schema": {
            "type": "object",
            "properties": {
                "path": {
                    "type": "string",
                    "description": "Path to the directory to list (default: current directory)",
                },
            },
        },
        "execute_fn": _list_directory,
    },
    "search_files": {
        "name": "search_files",
        "description": "Recursively search for files matching a glob pattern.",
        "input_schema": {
            "type": "object",
            "properties": {
                "path": {
                    "type": "string",
                    "description": "Directory to search in",
                },
                "pattern": {
                    "type": "string",
                    "description": "Glob pattern to match (e.g. '*.py', '**/*.json')",
                },
            },
            "required": ["path", "pattern"],
        },
        "execute_fn": _search_files,
    },
}


BUILTIN_REGISTRY = build_builtin_registry()
